### PR TITLE
Kja/controller

### PR DESCRIPTION
**Pull request by ktrops**

### CODE REVIEW

File: app/controllers/shipping_api_controller.rb [hunk 1]
```diff
@@ -1,8 +1,14 @@
...
+  before_action :set_shipping_rate, only: :rates
 
   def rates
-    render json: {}
+    if @shipping_rate.valid?
+      estimates = @shipping_rate.ups_rates + @shipping_rate.usps_rates
+      render json: estimates
+    else
+      render json: @shipping_rate.errors
+    end
```
Comment (catchingash) [inline app/controllers/shipping_api_controller.rb:12]: We probably want to add status codes to these.

File: app/controllers/shipping_api_controller.rb [hunk 2]
```diff
@@ -17,20 +23,41 @@ def create_log
...
+  end
+
+  def destination_params
+    params.require(:destination).permit(:city, :state, :zip, :country)
   end
 
-  def set_destination
+  def package
+    weight = package_params[:weight].to_i
+    dimensions = package_params[:dimensions].map{|d| d.to_i}
```
Comment (catchingash) [inline app/controllers/shipping_api_controller.rb]: Is `to_i` necessary on weight and dimensions? Or can the gem handle these as strings?

File: app/controllers/shipping_api_controller.rb [hunk 3]
```diff
@@ -17,20 +23,41 @@ def create_log
...
-    @shipping_rate = ShippingRate.new(@origin, @destination, @package)
+    @shipping_rate = ShippingRate.new(attributes = { origin: origin, destination: destination, package: package } )
   end
 
+<<<<<<< HEAD
+=======
   def log_params
     params.require(:log).permit(:customer, :order_id, :service, :cost, :origin, :destination)
   end
+>>>>>>> e7401120388bcc98350f8dba9e38603eff5fb5e1
```
Comment (catchingash) [inline app/controllers/shipping_api_controller.rb]: You'll want to handle the merge conflict here.

File: spec/controllers/shipping_api_controller_spec.rb [hunk 4]
```diff
@@ -24,13 +26,39 @@
       end
 
       describe "the returned json object" do
-        before :each do
-          @object = JSON.parse(response.body)
+        # before :each do
+        #   VCR.use_cassette("/rates", record: :new_episodes) do
+        #     post :rates, params
+        #   end
+        # end
```
Comment (catchingash) [inline spec/controllers/shipping_api_controller_spec.rb]: Looks like we can delete these lines?

File: spec/controllers/shipping_api_controller_spec.rb [hunk 5]
```diff
@@ -24,13 +26,39 @@
...
+
+        it "is an array of hashes" do
+          expect(object).to be_an_instance_of Array
+          expect(object.first).to be_an_instance_of Hash
+        end
+
+        it "includes only service, price, and delivery_date" do
+          object.each do |instance|
+            expect(instance.keys).to eq(["service", "price", "delivery_date"])
+          end
```
Comment (catchingash) [inline spec/controllers/shipping_api_controller_spec.rb]: Would you be willing to add another line to this one that checks to make sure that these are the ONLY keys included?

File: app/controllers/shipping_api_controller.rb (hunk 2 above)
Comment (ktrops) [inline app/controllers/shipping_api_controller.rb]: Yes, they have to be integers. The api can't handle strings

File: app/controllers/shipping_api_controller.rb [hunk 6]
```diff
@@ -1,8 +1,15 @@
...
+  skip_before_filter :verify_authenticity_token, only: [:rates, :create_log]
+  before_action :set_shipping_rate, only: :rates
 
   def rates
-    render json: {}
+    if @shipping_rate.valid?
+      estimates = @shipping_rate.ups_rates + @shipping_rate.usps_rates
+      render json: estimates
+    else
+      render json: @shipping_rate.errors
```
Comment (catchingash) [inline app/controllers/shipping_api_controller.rb]: We still need status codes on these, don't we?

**Pull request merged**